test/e2e/util.go

```diff
@@ -2298,6 +2298,60 @@ func waitForDeploymentStatus(c clientset.Interface, ns, deploymentName string, d
 	return err
 }
 
+// waitForDeploymentRollbackCleared waits for given deployment either started rolling back or doesn't need to rollback.
+func waitForDeploymentRollbackCleared(c clientset.Interface, ns, deploymentName string) error {
+	return wait.Poll(poll, 1*time.Minute, func() (bool, error) {
+		deployment, err := c.Extensions().Deployments(ns).Get(deploymentName)
+		if err != nil {
+			return false, err
+		}
+		// Rollback not set or is kicked off
+		if deployment.Spec.RollbackTo == nil {
+			return true, nil
+		}
+		return false, nil
+	})
+}
+
+// waitForDeploymentStatusUpdated waits for the deployment's status to match the given status fields. The input int* fields will be ignored if it's nil.
+func waitForDeploymentStatusUpdated(c clientset.Interface, ns, deploymentName string, replicas, updatedReplicas, availableReplicas, unavailableReplicas *int) error {
+	return wait.Poll(poll, 5*time.Minute, func() (bool, error) {
+		deployment, err := c.Extensions().Deployments(ns).Get(deploymentName)
+		if err != nil {
+			return false, err
+		}
+		if replicas != nil && deployment.Status.Replicas != *replicas ||
+			updatedReplicas != nil && deployment.Status.UpdatedReplicas != *updatedReplicas ||
+			availableReplicas != nil && deployment.Status.AvailableReplicas != *availableReplicas ||
+			unavailableReplicas != nil && deployment.Status.UnavailableReplicas != *unavailableReplicas {
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+// waitForDeploymentRevisionAndImage waits for the deployment's and its new RS's revision and container image to match the given revision and image.
+func waitForDeploymentRevisionAndImage(c clientset.Interface, ns, deploymentName string, revision, imageName, image string) error {
+	return wait.Poll(poll, 1*time.Minute, func() (bool, error) {
+		deployment, err := c.Extensions().Deployments(ns).Get(deploymentName)
+		if err != nil {
+			return false, err
+		}
+		newRS, err := deploymentutil.GetNewReplicaSet(deployment, c)
+		if err != nil {
+			return false, err
+		}
+		// Check revision of this deployment, and of the new replica set of this deployment
+		if deployment.Annotations == nil || deployment.Annotations[deploymentutil.RevisionAnnotation] != revision ||
+			newRS.Annotations == nil || newRS.Annotations[deploymentutil.RevisionAnnotation] != revision ||
+			deployment.Spec.Template.Spec.Containers[0].Name != imageName || deployment.Spec.Template.Spec.Containers[0].Image != image ||
+			newRS.Spec.Template.Spec.Containers[0].Name != imageName || newRS.Spec.Template.Spec.Containers[0].Image != image {
+			return false, nil
+		}
+		return true, nil
+	})
+}
```

Comment:
What kind of error does this produce upon timeout? Something that includes the deployment name?